Submissions/Ekey_Week 3 - Functions and Objects.playground/Contents.swift

```diff
@@ -0,0 +1,263 @@
+//}
+//greet("John", on: "Wednesday")
+
+// Laurie:  Not completely sure if this is what you are asking for but...
+func greet(_ person: String, eating lunchSpecial: String) -> String {
+    return "Hello \(person), todays lunch specal is \(lunchSpecial)."
+}
+greet("John", eating: "Fish and Chips")
+
+//: 2.)
+//:
+//: Rewrite the closure to return zero for all odd numbers.
+//:
+var numbers = [20, 19, 7, 12]
+
+//: full "syntactic sugar" form:
+//numbers.map({ (number: Int) -> Int in
+//    let result = 3 * number
+//    return result
+//})
+// Laurie:  Answers
+numbers.map({ (number: Int) -> Int in
+    let result = 3 * number
+//    let returnme = result % 2 == 0 ? result : 0
+    return result % 2 == 0 ? result : 0
+})
+//: omitting parameters and return type form:
+// let mappedNumbers = numbers.map({ number in 3 * number })
+// Laurie answer
+let mappedNumbers = numbers.map({ number in 3 * number % 2 == 0 ? 3 * number : 0 })
+print(mappedNumbers)
+//: reference variable form:
+//let mappedNums = numbers.map { 3 * $0 }
+// Laurie: answer
+let mappedNums = numbers.map { 3 * $0 % 2 == 0 ? 3 * $0 : 0  }
+print(mappedNums)
+
+
+//: ## Objects and Classes
+//: 3.)
+//:
+//: Add a constant property with `let`, and add another method that takes an argument.
+//:
+//: 4.)
+//:
+//: This version of the `Shape` class is missing something important: an initializer to set up the class when an instance is created. Use `init` to create one.
+//:
+//class Shape {
+//    var numberOfSides = 0
+//    func simpleDescription() -> String {
+//        return "A shape with \(numberOfSides) sides."
+//    }
+//}
+
+//Laurie:  Answer
+class Shape {
+    var numberOfSides = 0
+    let myVar = "Hello"
+    func simpleDescription() -> String {
+        return "A shape with \(numberOfSides) sides."
+    }
+    func takesArgument(_ greeting: String) {
+        print(greeting + " my friend")
+    }
+    init(count: Int) {
+        numberOfSides = count
+    }
+}
+
+let testLetwithClass = Shape(count: 5)
+testLetwithClass.takesArgument("Hola")
+testLetwithClass.numberOfSides = 6
+// cannot modify let property unless use mutable
+
+
+
+//: 5.)
+//:
+//: Make a subclass of `NamedShape` called `Circle` that takes a radius and a name as arguments to its initializer. Implement an `area()` and a `simpleDescription()` method on the `Circle` class.
+//:
+
+// Amanda -- I would change this to take just the radius, Circle itself should know what it is. i.e. a Circle -- so only 1 parm on the init/constructor
+
+class NamedShape {
+    var numberOfSides: Int = 0
+    var name: String
+    
+    init(name: String) {
+        self.name = name
+    }
+    
+    func simpleDescription() -> String {
+        return "A shape with \(numberOfSides) sides."
+    }
+}
+
+class Circle : NamedShape {
+    var radius: Double = 0
+    
+    // Laurie slight change... each object should know how to handle itself - shouldn't have to pass in "circle" from outside..
```

Comment:
And what if I want to call my circle "My kitty's favorite circle" to differentiate it from "That circle my kitty absolutely hates"? What if I have 400 circles my kitty absolutely hates and I need to have one circle named differently than all the others? :p You crack me up.